File: macos/Sources/Network/WireFrame.swift
import Foundation

public enum PacketType: UInt16, Sendable {
  case ping = 0x0001
  case pong = 0x0002
  case pairRequest = 0x0003
  case pairConfirm = 0x0004
  case deviceInfo = 0x0005
  case pairInvite = 0x0006
  case pairFail = 0x0007
  case pairAccept = 0x0008
  case clipText = 0x0010
  case clipUrl = 0x0011
  case clipImage = 0x0012
  case clipFile = 0x0013
  case fileStart = 0x0014
  case fileChunk = 0x0015
  case fileEnd = 0x0016
  case fileCancel = 0x0017
  case configSync = 0x0020
  case disconnect = 0x00FF
}

public struct WireFrame: Sendable, Equatable {
  public static let protocolVersion: UInt16 = 0x0001
  public static let headerSize: Int = 28
  public static let maxPayloadSize: Int = 104_857_600
  public static let authTagSize: Int = 16
  public static let nonceSize: Int = 12
  public static let streamingChunkSize: Int = 1_048_576
  public static let defaultPort: UInt16 = 58240
  public static let defaultClientPort: UInt16 = 58241
  public static let serviceType: String = "_clipsync._tcp"
  public static let clientServiceType: String = "_clipsync-cli._tcp"
  public let type: PacketType
  public let timestamp: UInt64
  public let nonce: Data
  public let ciphertext: Data
  public let tag: Data

  public init(
    type: PacketType, timestamp: UInt64 = UInt64(Date().timeIntervalSince1970 * 1000), nonce: Data,
    ciphertext: Data, tag: Data
  ) {
    self.type = type
    self.timestamp = timestamp
    self.nonce = nonce
    self.ciphertext = ciphertext
    self.tag = tag
  }

  public func serialize() -> Data {
    let payloadLength = UInt32(ciphertext.count + tag.count)
    var data = Data(capacity: Self.headerSize + ciphertext.count + tag.count)

    var lengthBig = payloadLength.bigEndian
    data.append(Data(bytes: &lengthBig, count: 4))

    var typeBig = type.rawValue.bigEndian
    data.append(Data(bytes: &typeBig, count: 2))

    var versionBig = Self.protocolVersion.bigEndian
    data.append(Data(bytes: &versionBig, count: 2))

    var timestampBig = timestamp.bigEndian
    data.append(Data(bytes: &timestampBig, count: 8))

    data.append(nonce.prefix(Self.nonceSize))
    data.append(ciphertext)
    data.append(tag.prefix(Self.authTagSize))

    return data
  }

  public static func deserialize(from data: Data) -> WireFrame? {
    guard data.count >= headerSize + authTagSize else {
      return nil
    }

    let parsedHeader: (UInt32, PacketType, UInt16, UInt64)? = data.withUnsafeBytes { ptr in
      let len = ptr.loadUnaligned(fromByteOffset: 0, as: UInt32.self).bigEndian
      let typeRaw = ptr.loadUnaligned(fromByteOffset: 4, as: UInt16.self).bigEndian
      let ver = ptr.loadUnaligned(fromByteOffset: 6, as: UInt16.self).bigEndian
      let time = ptr.loadUnaligned(fromByteOffset: 8, as: UInt64.self).bigEndian
      guard let type = PacketType(rawValue: typeRaw) else { return nil }
      return (len, type, ver, time)
    }

    guard let (payloadLength, packetType, version, timestamp) = parsedHeader,
      version == protocolVersion,
      Int(payloadLength) <= maxPayloadSize,
      data.count == headerSize + Int(payloadLength)
    else {
      return nil
    }
    let nonce = data.subdata(in: 16..<28)

    let ciphertextEnd = data.count - authTagSize
    guard ciphertextEnd >= headerSize else {
      return nil
    }

    let ciphertext = data.subdata(in: headerSize..<ciphertextEnd)
    let tag = data.subdata(in: ciphertextEnd..<data.count)

    return WireFrame(
      type: packetType,
      timestamp: timestamp,
      nonce: nonce,
      ciphertext: ciphertext,
      tag: tag
    )
  }
}

File: android/app/src/main/java/com/prabotics/rapidrop/network/WireFrame.kt
package com.prabotics.rapidrop.network

import java.nio.ByteBuffer
import java.nio.ByteOrder



enum class PacketType(val value: Short) {
    PING(0x0001),
    PONG(0x0002),
    PAIR_REQUEST(0x0003),
    PAIR_CONFIRM(0x0004),
    DEVICE_INFO(0x0005),
    PAIR_INVITE(0x0006),
    PAIR_FAIL(0x0007),
    PAIR_ACCEPT(0x0008),
    CLIP_TEXT(0x0010),
    CLIP_URL(0x0011),
    CLIP_IMAGE(0x0012),
    CLIP_FILE(0x0013),
    FILE_START(0x0014),
    FILE_CHUNK(0x0015),
    FILE_END(0x0016),
    FILE_CANCEL(0x0017),
    CONFIG_SYNC(0x0020),
    DISCONNECT(0x00FF.toShort());

    companion object {
        private val map = entries.associateBy { it.value }
        fun fromValue(value: Short): PacketType? = map[value]
    }
}

data class WireFrame(
    val type: PacketType,
    val timestamp: Long = System.currentTimeMillis(),
    val nonce: ByteArray,
    val ciphertext: ByteArray,
    val tag: ByteArray
) {
    companion object {
        const val PROTOCOL_VERSION: Short = 0x0001
        const val HEADER_SIZE: Int = 28
        const val MAX_PAYLOAD_SIZE: Int = 104_857_600
        const val AUTH_TAG_SIZE: Int = 16
        const val NONCE_SIZE: Int = 12
        const val STREAMING_CHUNK_SIZE: Int = 1_048_576
        const val DEFAULT_PORT: Int = 58240
        const val DEFAULT_CLIENT_PORT: Int = 58241
        const val SERVICE_TYPE: String = "_clipsync._tcp"
        const val CLIENT_SERVICE_TYPE: String = "_clipsync-cli._tcp"
        const val INACTIVITY_TIMEOUT_MS: Long = 16000L
        const val WATCHDOG_INTERVAL_MS: Long = 5000L
        const val HEARTBEAT_INTERVAL_MS: Long = 4000L
        const val PAIRING_INVITE_TIMEOUT_MS: Long = 30000L
        const val MAX_IN_FLIGHT_CHUNKS: Int = 4
        const val CHUNK_HEADER_SIZE: Int = 28
        fun deserialize(bytes: ByteArray): WireFrame? {
            if (bytes.size < HEADER_SIZE + AUTH_TAG_SIZE) return null

            val buffer = ByteBuffer.wrap(bytes).order(ByteOrder.BIG_ENDIAN)
            val payloadLength = buffer.int
            if (payloadLength < AUTH_TAG_SIZE || payloadLength > MAX_PAYLOAD_SIZE) return null
            if (bytes.size != HEADER_SIZE + payloadLength) return null

            val typeRaw = buffer.short
            val packetType = PacketType.fromValue(typeRaw) ?: return null

            val version = buffer.short
            if (version != PROTOCOL_VERSION) return null

            val timestamp = buffer.long

            val nonce = ByteArray(NONCE_SIZE)
            buffer.get(nonce)

            val ciphertextLength = payloadLength - AUTH_TAG_SIZE
            val ciphertext = ByteArray(ciphertextLength)
            buffer.get(ciphertext)

            val tag = ByteArray(AUTH_TAG_SIZE)
            buffer.get(tag)

            return WireFrame(
                type = packetType,
                timestamp = timestamp,
                nonce = nonce,
                ciphertext = ciphertext,
                tag = tag
            )
        }
    }

    fun serialize(): ByteArray {
        val payloadLength = ciphertext.size + tag.size
        val totalSize = HEADER_SIZE + payloadLength
        val buffer = ByteBuffer.allocate(totalSize).order(ByteOrder.BIG_ENDIAN)

        buffer.putInt(payloadLength)
        buffer.putShort(type.value)
        buffer.putShort(PROTOCOL_VERSION)
        buffer.putLong(timestamp)
        buffer.put(nonce, 0, minOf(nonce.size, NONCE_SIZE))
        buffer.put(ciphertext)
        buffer.put(tag, 0, minOf(tag.size, AUTH_TAG_SIZE))

        return buffer.array()
    }

    fun writeTo(out: java.io.OutputStream) {
        val payloadLength = ciphertext.size + tag.size
        val header = ByteArray(HEADER_SIZE)
        val buffer = ByteBuffer.wrap(header).order(ByteOrder.BIG_ENDIAN)
        buffer.putInt(payloadLength)
        buffer.putShort(type.value)
        buffer.putShort(PROTOCOL_VERSION)
        buffer.putLong(timestamp)
        buffer.put(nonce, 0, minOf(nonce.size, NONCE_SIZE))
        out.write(header, 0, HEADER_SIZE)
        out.write(ciphertext)
        out.write(tag, 0, minOf(tag.size, AUTH_TAG_SIZE))
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (javaClass != other?.javaClass) return false

        other as WireFrame
        if (type != other.type) return false
        if (timestamp != other.timestamp) return false
        if (!nonce.contentEquals(other.nonce)) return false
        if (!ciphertext.contentEquals(other.ciphertext)) return false
        if (!tag.contentEquals(other.tag)) return false

        return true
    }

    override fun hashCode(): Int {
        var result = type.hashCode()
        result = 31 * result + timestamp.hashCode()
        result = 31 * result + nonce.contentHashCode()
        result = 31 * result + ciphertext.contentHashCode()
        result = 31 * result + tag.contentHashCode()
        return result
    }
}
